### PR TITLE
Add Squash merge

### PR DESCRIPTION
**Pull request by noahfraiture**
- **PR Description**

Hello,

This PR add merge --squash. A PR already exist https://github.com/jesseduffield/lazygit/pull/3130, but the author abandoned it, so I remake it. 
I modified to fit most of the comment made except this one https://github.com/jesseduffield/lazygit/pull/3130/files#r1404808121. I didn't find an existing example and thus didn't know to modify the code to fit it to the new way of doing things. 

There's still the choice box to commit or not to do as discussed https://github.com/jesseduffield/lazygit/pull/3130#issuecomment-2112324990. I'll do it when I have time, I first need to read the code to see how it really works.

Also only english has been made for now. 

image

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

### CODE REVIEW

**Comment by noahfraiture** [comment]
Hello, I finished the feature, and it seems to work well, I will add tests in futures days, but I would like your feedback already if you feel it's needed. The only problem is to generate the cheatsheet, they seem to be empty for the squash, but I don't find where the generate takes them from. It's my first PR, hope it's good

File: pkg/gui/controllers/branches_controller.go
Comment (AzraelSec) [inline pkg/gui/controllers/branches_controller.go:122]: What about using `GetDisabledReson` here? This operation does not make sense with multiple items/multiple items handling is not implemented yet.
```suggestion
			Handler:     opts.Guards.OutsideFilterMode(self.squash),
			GetDisabledReason: self.require(self.singleItemSelected()),
```

**Review by AzraelSec — COMMENTED** [review]
Hello 👋🏻 I left a couple of comments on your PR to discuss while you write the tests.

**Comment by noahfraiture** [comment]
I have your suggestion which seems indeed better suited but there's a bug and I can't figure out if it's mine or not. The feature is always disabled even if the ShowErrorInPanel is set to false.
image
I wrote the bool value for debugging purpose and you can see that it is false and the feature should be enabled.

Do you have any idea why ?

**Review by AzraelSec — COMMENTED** [review]
@noahfraiture - The [ShowErrorInPanel](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/types/common.go#L212-L216) function just changes how lazygit displays an error when the user tries to run a disabled action. The DisabledReason field must be nil to make an action available, so you can conditionally assign your structure to it when the required conditions for that command are met.

That said, I'd suggest updating GetDisabledReason to disable the menu in the case of merging the checked-out branch, instead of disabling the individual menu options. This approach is similar to how we handle rebase operations. Here is an example of what I would do: https://github.com/jesseduffield/lazygit/commit/c36bbb10cfdaeb755b7429d5d604cd978fc8de6c.

Also, I think it would be better to clean up the history of this branch before proceeding to merge it.

**Comment by noahfraiture** [comment]
Thank you for your feedback. I applied your advices and made better tests. I also cleaned the commit if it's ok for you like this

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (AzraelSec) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:350]:
```suggestion
						"checkedOutBranch": checkedOutBranchName,
```

File: pkg/integration/tests/branch/squash.go
Comment (AzraelSec) [inline pkg/integration/tests/branch/squash.go:33]:
```suggestion
				Contains("original-branch").IsSelected(),
```

**Review by AzraelSec — COMMENTED** [review]
Nice work, @noahfraiture! 👍🏻  I left two small suggestions (sorry, I noticed them only now), but the PR looks good. After addressing these last comments, I'll approve your PR so that one of the maintainers can take a look themselves.

**Comment by noahfraiture** [comment]
> Nice work, @noahfraiture! 👍🏻 I left two small suggestions (sorry, I noticed them only now), but the PR looks good. After addressing these last comments, I'll approve your PR so that one of the maintainers can take a look themselves.

This is good suggestion, I should have made them in the first place. Done and tested

**Review by AzraelSec — COMMENTED** [review]
Thanks @noahfraiture, looks good to me 💪🏻 I would have suggested you to fix up the previous commits instead of adding a new one, but let's see what the maintainers think. Thanks for contributing!

@stefanhaller @jesseduffield this PR looks good to me - feel free to look at this when you have space for it 🙏🏻

**Comment by noahfraiture** [comment]
Hello, is there something more that I can do for this PR ?

File: pkg/commands/git_commands/commit.go
Comment (jesseduffield) [inline pkg/commands/git_commands/commit.go:319]:
```suggestion
		Arg("-m", fmt.Sprintf("Squash merge branch %s into %s", selectedBranch, checkedOutBranch)).
```

File: pkg/gui/controllers/remote_branches_controller.go
Comment (jesseduffield) [inline pkg/gui/controllers/remote_branches_controller.go:201]: This check is not necessary: it's impossible to merge into yourself if you're merging a remote branch

File: pkg/commands/git_commands/commit.go
Comment (jesseduffield) [inline pkg/commands/git_commands/commit.go:317]:
```suggestion
func (self *CommitCommands) CreateSquashMergeCommit(selectedBranch, checkedOutBranch string) error {
```

File: pkg/config/user_config.go
Comment (jesseduffield) [inline pkg/config/user_config.go:430]:
```suggestion
	SquashMerge           string `yaml:"squashMerge"`
```

File: pkg/gui/controllers/branches_controller.go
Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:607]:
```suggestion
func (self *BranchesController) squashMerge() error {
```

File: pkg/gui/controllers/branches_controller.go
Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:124]: Let's add `OpensMenu: true` to this so that the user can see that the keybinding opens a menu

Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:124]:
```suggestion
			Description:       self.c.Tr.SquashMerge,
```

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (jesseduffield) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:327]:
```suggestion
func (self *MergeAndRebaseHelper) SquashMerge(refName string) error {
```

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (jesseduffield) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:330]:
```suggestion
		Title: self.c.Tr.SquashMergeTitle,
```

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (jesseduffield) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:359]:
```suggestion
func (self *MergeAndRebaseHelper) SquashMergeUncommitted(refName string) func() error {
```

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (jesseduffield) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:367]:
```suggestion
func (self *MergeAndRebaseHelper) SquashMergeCommitted(refName, checkedOutBranchName string) func() error {
```

File: pkg/integration/tests/branch/squash.go
Comment (jesseduffield) [inline pkg/integration/tests/branch/squash.go:9]:
```suggestion
	Description:  "Squash merge a branch without committing",
```

**Review by jesseduffield — CHANGES_REQUESTED** [review]
Looking good, left some feedback.

The main thing is wording: We need to use `squash merge` and not `squash branch` or `squash files` because the former is less confusing, and you may be squash merging something other than a branch (e.g. a tag).

I also prefer @stefanhaller 's original wording on the menu items i.e. 'Squash merge and leave uncommitted' and 'Squash merge and commit'.

Also, what currently happens if you choose to squash merge and commit, but there are merge conflicts? I suspect after conflicts are resolved, lazygit won't attempt to then create the merge commit. It's easier when git itself knows it's trying to merge something, so lazygit can prompt to continue to the merge, but does this also happen with squash merge?

**Comment by noahfraiture** [comment]
> Looking good, left some feedback.
> 
> The main thing is wording: We need to use `squash merge` and not `squash branch` or `squash files` because the former is less confusing, and you may be squash merging something other than a branch (e.g. a tag).
> 
> I also prefer @stefanhaller 's original wording on the menu items i.e. 'Squash merge and leave uncommitted' and 'Squash merge and commit'.

Thank you for your feedback, I added your suggestions and adapt the remaining mention of squash to fit the new wording.

File: docs/Config.md
Comment (jesseduffield) [inline docs/Config.md:474]:
```suggestion
    squashMerge: S
```

File: pkg/gui/controllers/branches_controller.go
Comment (jesseduffield) [reply pkg/gui/controllers/branches_controller.go:124]: ^ missed this one

File: pkg/gui/controllers/helpers/merge_and_rebase_helper.go
Comment (jesseduffield) [inline pkg/gui/controllers/helpers/merge_and_rebase_helper.go:361]:
```suggestion
		self.c.LogAction(self.c.Tr.Actions.SquashMerge)
```

File: pkg/gui/controllers/remote_branches_controller.go
Comment (jesseduffield) [inline pkg/gui/controllers/remote_branches_controller.go:62]:
```suggestion
			Description: self.c.Tr.SquashMerge,
```

File: pkg/i18n/english.go
Comment (jesseduffield) [inline pkg/i18n/english.go:1766]:
```suggestion
			SquashMerge:                   "Squash merge",
```

File: pkg/integration/tests/branch/squash.go
Comment (jesseduffield) [inline pkg/integration/tests/branch/squash.go:9]:
```suggestion
	Description:  "Squash merge a branch both with and without committing",
```

File: pkg/integration/tests/branch/squash.go
Comment (jesseduffield) [inline pkg/integration/tests/branch/squash.go:52]:
```suggestion
			NavigateToLine(Contains("change-worktree-branch")).
```
Just a bit clearer

File: pkg/integration/tests/test_list.go
Comment (jesseduffield) [inline pkg/integration/tests/test_list.go:62]:
```suggestion
	branch.SquashMerge,
```

**Review by jesseduffield — CHANGES_REQUESTED** [review]
Just a few more things

**Comment by noahfraiture** [comment]
Done but a modification changed the translation file that already had "Squash" I adapt for every language to "SquashMerge" but I don't know if it's what's wanted

File: pkg/i18n/english.go
Comment (jesseduffield) [inline pkg/i18n/english.go:1097]: Sorry: we should keep the original Squash item here and add a new SquashMerge item. Squashing and squash merging are two separate things and we do not want to conflate them

**Review by jesseduffield — CHANGES_REQUESTED** [review]
one more thing

File: pkg/i18n/english.go
Comment (noahfraiture) [reply pkg/i18n/english.go:1097]: ok done

File: pkg/i18n/korean.go
Comment (jesseduffield) [inline pkg/i18n/korean.go:91]: I think we should just leave non-english translation files alone: This particular example is not particularly useful.

**Review by jesseduffield — CHANGES_REQUESTED** [review]
Sorry another thing: let's revert the changes to non-english i18n files. We're going to have software for doing this with AI very soon so no need to do it ourselves here (especially given we're not very good translators)

File: pkg/i18n/korean.go
Comment (noahfraiture) [reply pkg/i18n/korean.go]: bruh, ok done

**Comment by jesseduffield** [comment]
I've pushed some changes: I've replaced the merge confirmation popup with a menu which includes the option to squash merge:

image

This has a few benefits:
1) anywhere you can merge something, you can now also squash merge it (no need to add logic for each view)
2) logically it makes sense to have squash merge appearing in a menu alongside a logical merge
3) it allows us to use the 'S' keybinding for something else.

What do people think? @stefanhaller I'm keen to get your thoughts on this

**Comment by stefanhaller** [comment]
Yes, this makes a lot of sense. So much that I'm actually surprised nobody has come up with the idea before, it seems so obvious in hindsight.

Will test and review later today.

File: pkg/gui/controllers/branches_controller.go
Comment (stefanhaller) [inline pkg/gui/controllers/branches_controller.go:640]: This looks unused.

File: pkg/i18n/translations/ja.json
Comment (stefanhaller) [inline pkg/i18n/translations/ja.json:13]: What's going on here, did you edit these manually? I don't think we should do that, it's error-prone and unnecessary. The source of truth for these files should be crowdin (once we have it), until then we should simply leave these untouched.

**Review by stefanhaller — COMMENTED** [review]
Awesome, I like this.

I pushed one fixup with just very minor tweaks (7dc061d45955), and a bigger one (482e9674f67a) that makes the commit message configurable (I really like to use `[SQUASHED] branch-name` myself, so that my squash merges stand out more in the log).

This second one also removes the `CreateSquashMergeCommit` function that we no longer need now. The difference to before is that for people who have `git.commit.signOff` set to true, they now get a sign-off added to their squash merge commits, which makes sense I guess (although we don't do it for regular merges).

Two more things inline below.

File: pkg/i18n/english.go
Comment (stefanhaller) [inline pkg/i18n/english.go:1283]: "current working tree" reads a bit strange, as if there were several. I'd probably just drop the word "current".

File: pkg/i18n/translations/ja.json
Comment (jesseduffield) [reply pkg/i18n/translations/ja.json:13]: Indeed: I edited them manually. If crowdin is going to be the new source of truth for these translation files, I'm happy to leave them alone.

I've just set up a crowdin account and sent you an invite. I've also applied for the open-source licence so that it's free

**Comment by jesseduffield** [comment]
I've addressed your comments @stefanhaller so I'm mergin

Thanks for raising this PR @noahfraiture !

**Pull request merged**
